### PR TITLE
IK Test with a variety of constraints

### PR DESCRIPTION
**Pull request by mauricefallon**
I wanted to explore using the IK solver directly through C++
This adds an IK Test with a variety of constraints. There is already testIK.cpp but it doesn't really do much.

Also adds a bug fix for the original test.

Question:
Is the matlab IK solver (as used in the director etc) calling this C++ API via a wrapper. Or are there constraints which exist in matlab and not in C++ and vice-versa?

Attached is the IK result for the problem I added

screenshot-2015-10-06_18 40 07

### CODE REVIEW

**Comment by RussTedrake** [comment]
Short story:  the +drakeFunction classes have not been moved to c++ yet, but the RigidBodyConstraint classes have.  The traditional IK interface uses the RigidBodyConstraint classes, but our newer planners use both.  You'll be able to get pretty far using just the c++ IK.  And it should be getting better fast.

File: drake/systems/plants/test/testIKMoreConstraints.cpp [hunk 1]
```diff
@@ -0,0 +1,221 @@
+#include <drake/RigidBodyIK.h>
+#include <drake/RigidBodyManipulator.h>
+#include <drake/RigidBodyConstraint.h>
+
+#include <drake/IKoptions.h>
+#include <iostream>
+#include <cstdlib>
+#include <limits>
+#include <boost/shared_ptr.hpp>
```
Comment (tkoolen) [inline drake/systems/plants/test/testIKMoreConstraints.cpp]: Boost is not a dependency, hence the failure on windows. shared_ptr is std nowadays, just `#include <memory>`

File: drake/systems/plants/test/testIKMoreConstraints.cpp [hunk 2]
```diff
@@ -0,0 +1,221 @@
+#include <drake/RigidBodyIK.h>
+#include <drake/RigidBodyManipulator.h>
+#include <drake/RigidBodyConstraint.h>
+
+#include <drake/IKoptions.h>
```
Comment (tkoolen) [inline drake/systems/plants/test/testIKMoreConstraints.cpp]: These are going to refer to the headers in build/include. This is not what we want. They should be relative to the current directory, or in folders in drake-distro/drake included using CMake's include_directories.

File: drake/systems/plants/test/testIKMoreConstraints.cpp (hunk 1 above)
Comment (tkoolen) [inline drake/systems/plants/test/testIKMoreConstraints.cpp]: another boost shared_ptr.

File: drake/systems/plants/test/testIKMoreConstraints.cpp [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+int main()
+{
+
+  RigidBodyManipulator rbm("examples/Atlas/urdf/atlas_minimal_contact.urdf");
+  RigidBodyManipulator* model = &rbm;
+
+
+  int r_hand, l_foot, r_foot;
+  for(int i = 0;i<model->bodies.size();i++)
```
Comment (tkoolen) [inline drake/systems/plants/test/testIKMoreConstraints.cpp]: Just use the findLink function of RigidBodyManipulator.

File: drake/systems/plants/test/testIKMoreConstraints.cpp [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+      l_foot = i;
+    }
+    if(model->bodies[i]->linkname.compare(string("r_foot")) ==0 )
+    {
+      r_foot = i;
+    }
+  }
+
+
+  if(!model)
```
Comment (tkoolen) [inline drake/systems/plants/test/testIKMoreConstraints.cpp]: This is impossible.

File: drake/systems/plants/test/testIKMoreConstraints.cpp [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+  // 1 CoM Constraint
+  Vector3d com_lb = Vector3d::Zero(); 
+  Vector3d com_ub = Vector3d::Zero(); 
+  com_lb(0) = -0.1;
+  com_lb(1) = -0.1;
+  com_lb(2) = 0.6;
+  com_ub(0) = 0.1;
+  com_ub(1) = 0.1;
+  com_ub(2) = 1.0;
+  WorldCoMConstraint* kc_com = new WorldCoMConstraint(model,com_lb,com_ub,tspan);
```
Comment (tkoolen) [inline drake/systems/plants/test/testIKMoreConstraints.cpp]: Don't use raw pointers. If you must use a pointer, use unique_ptr in this case, but better yet, just use
```
WorldCoMConstraint kc_com(model,com_lb,com_ub,tspan);
```

File: drake/systems/plants/test/testIKMoreConstraints.cpp [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+
+  // 2 Back Posture Constraint
+  PostureConstraint* kc_posture_back = new PostureConstraint(model, tspan);
```
Comment (tkoolen) [inline drake/systems/plants/test/testIKMoreConstraints.cpp]: raw pointer

File: drake/systems/plants/test/testIKMoreConstraints.cpp [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+  std::vector<int> back_idx;
+  back_idx.push_back(6);
+  back_idx.push_back(7);
+  back_idx.push_back(8);
+  VectorXd back_lb = VectorXd::Zero(3);
+  VectorXd back_ub = VectorXd::Zero(3);
+  kc_posture_back->setJointLimits(3, back_idx.data(), back_lb, back_ub);
+
+  // 3 Knees Constraint
+  PostureConstraint* kc_posture_knees = new PostureConstraint(model, tspan);
```
Comment (tkoolen) [inline drake/systems/plants/test/testIKMoreConstraints.cpp]: raw pointer

File: drake/systems/plants/test/testIKMoreConstraints.cpp [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+  rhand_pos0(2) = 0.4;
+
+  Vector3d rhand_pos_lb = rhand_pos0;
+  rhand_pos_lb(0) +=0.05;
+  rhand_pos_lb(1) +=0.05;
+  rhand_pos_lb(2) +=0.05;
+  Vector3d rhand_pos_ub = rhand_pos_lb;
+  rhand_pos_ub(2) += 0.05;
+  //std::cout << rhand_pos_ub.transpose() << " rhand\n" ;
+  WorldPositionConstraint* kc_rhand = new WorldPositionConstraint(model,r_hand,r_hand_pt,rhand_pos_lb,rhand_pos_ub,tspan);
```
Comment (tkoolen) [inline drake/systems/plants/test/testIKMoreConstraints.cpp]: raw pointer

File: drake/systems/plants/test/testIKMoreConstraints.cpp [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+
+
+  // 7 QuasiStatic Constraints
+  QuasiStaticConstraint* kc_quasi = new QuasiStaticConstraint(model, tspan);
```
Comment (tkoolen) [inline drake/systems/plants/test/testIKMoreConstraints.cpp]: raw pointer

File: drake/systems/plants/test/testIKMoreConstraints.cpp [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+  lfoot_pos0(1) = 0.13;
+  lfoot_pos0(2) = 0.08;
+  Vector3d lfoot_pos_lb = lfoot_pos0;
+  lfoot_pos_lb(0) +=0.001;
+  lfoot_pos_lb(1) +=0.001;
+  lfoot_pos_lb(2) +=0.001;
+  Vector3d lfoot_pos_ub = lfoot_pos_lb;
+  lfoot_pos_ub(2) += 0.01;
+  // std::cout << lfoot_pos0.transpose() << " lfoot\n" ;
+  WorldPositionConstraint* kc_lfoot_pos = new WorldPositionConstraint(model,l_foot,l_foot_pt,lfoot_pos_lb,lfoot_pos_ub,tspan);
```
Comment (tkoolen) [inline drake/systems/plants/test/testIKMoreConstraints.cpp]: raw pointer

File: drake/systems/plants/test/testIKMoreConstraints.cpp [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+  Eigen::Vector4d quat_des(1,0,0,0);
+  double tol = 0.0017453292519943296;
+  WorldQuatConstraint* kc_lfoot_quat = new  WorldQuatConstraint(model, l_foot, quat_des, tol, tspan);
```
Comment (tkoolen) [inline drake/systems/plants/test/testIKMoreConstraints.cpp]: raw pointer

File: drake/systems/plants/test/testIKMoreConstraints.cpp [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+  rfoot_pos0(1) = -0.13;
+  rfoot_pos0(2) = 0.08;
+  Vector3d rfoot_pos_lb = rfoot_pos0;
+  rfoot_pos_lb(0) +=0.001;
+  rfoot_pos_lb(1) +=0.001;
+  rfoot_pos_lb(2) +=0.001;
+  Vector3d rfoot_pos_ub = rfoot_pos_lb;
+  rfoot_pos_ub(2) += 0.001;
+  //std::cout << rfoot_pos0.transpose() << " lfoot\n" ;
+  WorldPositionConstraint* kc_rfoot_pos = new WorldPositionConstraint(model,r_foot,r_foot_pt,rfoot_pos_lb,rfoot_pos_ub,tspan);
```
Comment (tkoolen) [inline drake/systems/plants/test/testIKMoreConstraints.cpp]: raw pointer

File: drake/systems/plants/test/testIKMoreConstraints.cpp [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+  WorldQuatConstraint* kc_rfoot_quat = new  WorldQuatConstraint(model, r_foot, quat_des, tol, tspan);
```
Comment (tkoolen) [inline drake/systems/plants/test/testIKMoreConstraints.cpp]: raw pointer

File: drake/systems/plants/test/testIKMoreConstraints.cpp [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+  //std::cout << l_foot_pts <<"\n";
+  kc_quasi->addContact(1, &l_foot, &l_foot_pts);
+  Eigen::Matrix3Xd r_foot_pts = Eigen::Matrix3Xd::Zero(3, 4);
+  r_foot_pts <<    -0.0820 ,  -0.0820 ,   0.1780 ,   0.1780   , 0.0624  , -0.0624  ,  0.0624 ,  -0.0624  , -0.0811,   -0.0811  , -0.0811 ,  -0.0811;
+  kc_quasi->addContact(1, &r_foot, &r_foot_pts);
+
+
+
+  int num_constraints = 9;
+  RigidBodyConstraint** constraint_array = new RigidBodyConstraint*[num_constraints];
```
Comment (tkoolen) [inline drake/systems/plants/test/testIKMoreConstraints.cpp]: just use a std::vector, and then pass vec.data() into inverseKin (until we clean up the inverseKin interface to take a const std::vector&).

**Comment by tkoolen** [comment]
The test makes a lot of assumptions about the indices of joints. It would be much easier to understand and maintain if you used RigidBodyManipulator::findLink, the velocity_num_start member of RigidBody, and the getNumVelocities method of DrakeJoint. Consider what would happen if we switched to using a quaternion parameterized floating joint.

**Comment by mauricefallon** [comment]
Yikes! Mea culpa. ;)

I've addressed the includes and the use of pointers/vectors. The later was used in testIK.cpp

>  It would be much easier to understand and maintain if you used RigidBodyManipulator::findLink, the 
> velocity_num_start member of RigidBody, and the getNumVelocities method of DrakeJoint.

Could you give me an example of finding a joint ID?
findJointId('back_bkz') does correspond to looping through the positions

for (int i=0 ; i <model->num_positions ; i++){
    std::cout << model->getPositionName(i) << " " << i << "\n";
}

To be honest I don't understand the difference between positions and joints.

**Comment by tkoolen** [comment]
Thanks and no worries. We just want to make sure that this code is clean, because tests also serve the purpose of being examples for users.

A joint may have more than one degree of freedom. Also, the number of positions may not be the same as the number of velocities (e.g. a quaternion parameterized floating joint has 7 position variables whereas the velocity is the twist, a 6D vector), but that's a different matter.

In your example, back_bkz happens to be the second joint in the tree of rigid bodies/joints, but it is preceded by an rpy parameterized floating joint with 6 position variables (indexed 0 through 5). Hence model.getPositionName(6) returns back_bkz.

I would do something like this. Create a function:
```c++
#include <numeric> // for iota
vector<int> jointPositionVectorIndices(const RigidBodyManipulator& model, const std::string& name) {
  shared_ptr<RigidBody> joint_parent_body = model.findJoint(name);
  int num_positions = joint_parent_body->getJoint().getNumPositions();
  vector<int> ret(static_cast<size_t>(num_positions));

  // fill with sequentially increasing values, starting at joint_parent_body->position_num_start:
  iota(ret.begin(), ret.end(), joint_parent_body->position_num_start);
  return ret;
}
```
which you can call as follows:
```c++
auto position_indices = jointPositionVectorIndices(model, "base");
for (int i: position_indices) std::cout << i << ' ';
std::cout << endl;

position_indices = jointPositionVectorIndices(model, "l_leg_kny");
for (int i: position_indices) std::cout << i << ' ';
```
and outputs
```
0 1 2 3 4 5 
13 
```

**Comment by mauricefallon** [comment]
Thanks for the background info @tkoolen , its appreciated. Clears up my understanding of the floating base joint.

I think the latest commit covers your suggestions. Perhaps you also want to remove testIK.cpp as this test is more complete

**Comment by RussTedrake** [comment]
somehow my build logic failed me (it emailed me the results instead of posting here):

*** BUILD FAILED with compilation errors. ***

Site: drake005
Build Name: mitdrc-mfallon-ik-constraint-exmaple
Build Server Directory: /private/tmp/drake-23470

Most recent commits to drake included in this build:
```
5475254 use suggested approach to find joint angles, remove hard coding
afce95b removes installed include, uses std::vector, removes shared pointer
dc6389a rebase fix
8f0b70a little fix for pitch when height was meant
1055b74 adding Ik test with a variety of constraints
```

```
470/build/include/bullet/LinearMath/btList.h
-- Installing:
/tmp/drake-23470/build/include/bullet/LinearMath/btMatrix3x3.h
-- Installing: /tmp/drake-23470/build/include/bullet/LinearMath/btMatrixX.h
-- Installing: /tmp/drake-23470/build/include/bullet/LinearMath/btMinMax.h
-- Installing:
/tmp/drake-23470/build/include/bullet/LinearMath/btMotionState.h
-- Installing:
/tmp/drake-23470/build/include/bullet/LinearMath/btPolarDecomposition.h
-- Installing:
/tmp/drake-23470/build/include/bullet/LinearMath/btPoolAllocator.h
-- Installing:
/tmp/drake-23470/build/include/bullet/LinearMath/btQuadWord.h
-- Installing:
/tmp/drake-23470/build/include/bullet/LinearMath/btQuaternion.h
-- Installing:
/tmp/drake-23470/build/include/bullet/LinearMath/btQuickprof.h
-- Installing: /tmp/drake-23470/build/include/bullet/LinearMath/btRandom.h
-- Installing: /tmp/drake-23470/build/include/bullet/LinearMath/btScalar.h
-- Installing:
/tmp/drake-23470/build/include/bullet/LinearMath/btSerializer.h
-- Installing:
/tmp/drake-23470/build/include/bullet/LinearMath/btSpatialAlgebra.h
-- Installing:
/tmp/drake-23470/build/include/bullet/LinearMath/btStackAlloc.h
-- Installing:
/tmp/drake-23470/build/include/bullet/LinearMath/btTransform.h
-- Installing:
/tmp/drake-23470/build/include/bullet/LinearMath/btTransformUtil.h
-- Installing: /tmp/drake-23470/build/include/bullet/LinearMath/btVector3.h
[100%] Completed 'bullet3'
[100%] Built target bullet3
Scanning dependencies of target fixup_rpath_libBullet2FileLoader.2.83.dylib
Scanning dependencies of target fixup_rpath_libLinearMath.2.83.dylib
Scanning dependencies of target fixup_rpath_libBullet3Collision.2.83.dylib
Scanning dependencies of target fixup_rpath_libBullet3Dynamics.2.83.dylib
Scanning dependencies of target fixup_rpath_libBulletCollision.2.83.dylib
Scanning dependencies of target
fixup_rpath_libBullet3OpenCL_clew.2.83.dylib
Scanning dependencies of target fixup_rpath_libBulletSoftBody.2.83.dylib
Scanning dependencies of target fixup_rpath_libBullet3Geometry.2.83.dylib
Scanning dependencies of target fixup_rpath_libBullet3Common.2.83.dylib
Scanning dependencies of target fixup_rpath_libBulletDynamics.2.83.dylib
[100%] [100%] [100%] [100%] Built target
fixup_rpath_libBullet2FileLoader.2.83.dylib
[100%] Built target fixup_rpath_libBullet3Geometry.2.83.dylib
Built target fixup_rpath_libBullet3Collision.2.83.dylib
[100%] Built target fixup_rpath_libBullet3Dynamics.2.83.dylib
[100%] Built target fixup_rpath_libBullet3OpenCL_clew.2.83.dylib
[100%] Built target fixup_rpath_libBulletDynamics.2.83.dylib
[100%] Built target fixup_rpath_libBullet3Common.2.83.dylib
Built target fixup_rpath_libBulletSoftBody.2.83.dylib
[100%] Built target fixup_rpath_libBulletCollision.2.83.dylib
Built target fixup_rpath_libLinearMath.2.83.dylib
Scanning dependencies of target install
[100%] Built target install
[ 88%] No install step for 'bullet'
[ 88%] Completed 'bullet'
[ 88%] Built target bullet
make[1]: *** [all] Error 2
make: *** [all] Error 2
```

**Comment by tkoolen** [comment]
Thanks for the cleanup, looks good to me now!
I actually can't reproduce the build failure myself (on OSX, like drake005, however only checking out the branch, not merging it into master). The actual error is also missing from the log above. Maybe close the pull request and reopen to re-trigger the build servers?

**Comment by mauricefallon** [comment]
Great, I'm going to leave it to you both to resolve the build servers. This branch is based off a week old version of master.

**Comment by tkoolen** [comment]
Good to go?

**Comment by rdeits** [comment]
Only Travis has actually tested this pull request. It probably needs to get closed and re-opened.

**Comment by tkoolen** [comment]
@mauricefallon or @RussTedrake, could you please close and reopen?

**Comment by tkoolen** [comment]
I can fix this pull request, but I'd prefer to do it after https://github.com/RobotLocomotion/drake/pull/1434 is in to prevent double work.

**Comment by tkoolen** [comment]
@RussTedrake, could you trigger the other build servers for this pull request as well?

**Pull request merged**